File: R/ssa_simulation.R
# boilerplate function but allows for parameter name auto completion
#' @importFrom methods new
create_simulation <- function(
  compiled_reactions,
  params = NULL,
  method_ptr = ssa_exact()$factory(),
  initial_state,
  census_interval = 0,
  log_propensity = FALSE,
  log_firings = FALSE,
  log_buffer = FALSE,
  stop_on_neg_state = TRUE,
  final_time,
  max_walltime = Inf,
  sim_name = NA_character_,
  verbose = FALSE,
  console_interval = 1
) {
  # create new simulation object
  sim <- new(SSA_simulation)

  # set propensity functions
  sim$initialise(
    compiled_reactions$num_functions,
    compiled_reactions$functions_pointer,
    params,
    compiled_reactions$buffer_size,
    method_ptr,
    initial_state,
    compiled_reactions$state_change@i,
    compiled_reactions$state_change@p,
    compiled_reactions$state_change@x,
    census_interval,
    log_propensity,
    log_firings,
    log_buffer,
    stop_on_neg_state,
    final_time,
    max_walltime,
    sim_name,
    verbose,
    console_interval
  )
  # make sure these pointers do not get GCed
  attr(sim, "prop_ptr") <- compiled_reactions$functions_pointer
  attr(sim, "method_ptr") <- method_ptr

  sim$reset()

  sim
}


#' Invoking the stochastic simulation algorithm
#'
#' Main interface function to the implemented \acronym{SSA} methods. Runs a
#' single realization of a predefined system. For a detailed explanation
#' on how to set up your first \acronym{SSA} system, check the introduction
#' vignette: `vignette("an_introduction", package = "GillespieSSA2")`.
#' If you're transitioning from \pkg{GillespieSSA} to \pkg{GillespieSSA2},
#' check out the corresponding vignette:
#' `vignette("converting_from_GillespieSSA", package = "GillespieSSA2")`.
#'
#' Substantial improvements in speed and accuracy can be obtained by
#' adjusting the additional (and optional) `ssa` arguments. By default
#' `ssa` uses conservative parameters (o.a. [ssa_exact()]) which prioritise
#' computational accuracy over computational speed.
#'
#' Approximate methods ([ssa_etl()] and [ssa_btl()]) are not fool proof!
#' Some tweaking might be required for a stochastic model to run appropriately.
#'
#' @param initial_state `[named numeric vector]` The initial state to start the simulation with.
#' @param reactions A list of reactions, see [reaction()].
#' @param final_time `[numeric]` The final simulation time.
#' @param params `[named numeric vector]` Constant parameters to be used in the propensity functions.
#' @param method `[ssa_method]`] Which SSA algorithm to use. Must be one of: [ssa_exact()],
#'   [ssa_btl()], or [ssa_etl()].
#' @param census_interval `[numeric]` The approximate interval between recording the state of the system.
#'   Setting this parameter to `0` will cause each state to be recorded, and
#'   to `Inf` will cause only the end state to be recorded.
#' @param stop_on_neg_state `[logical]` Whether or not to stop the simulation when
#'   the a negative value in the state has occured. This can occur, for instance, in the [ssa_etl()]
#'   method.
#' @param max_walltime `[numeric]` The maximum duration (in seconds) that the
#'   simulation is allowed to run for before terminated.
#' @param log_propensity `[logical]` Whether or not to store the propensity values at each census.
#' @param log_firings `[logical]` Whether or not to store number of firings of each reaction between censuses.
#' @param log_buffer `[logical]` Whether or not to store the buffer at each census.
#' @param verbose `[logical]` If `TRUE`, intermediary information pertaining to the simulation will be displayed.
#' @param console_interval `[numeric]` The approximate interval between intermediary information outputs.
#' @param sim_name `[character]` An optional name for the simulation.
#'
#' @seealso [GillespieSSA2] for a high level explanation of the package
#'
#' @return Returns a list containing the output of the simulation:
#'
#' * `out[["time"]]`: `[numeric]` The simulation time at which a census was performed.
#' * `out[["state"]]`: `[numeric matrix]` The number of individuals at those time points.
#' * `out[["propensity"]]`: `[numeric matrix]` If `log_propensity` is `TRUE`, the propensity value of each reaction at each time point.
#' * `out[["firings"]]`: `[numeric matrix]` If `log_firings` is `TRUE`, the number of firings between two time points.
#' * `out[["buffer"]]`: `[numeric matrix]` If `log_buffer` is `TRUE`, the buffer values at each time point.
#' * `out[["stats"]]`: `[data frame]` Various stats:
#'   - `$method`: The name of the SSA method used.
#'   - `$sim_name`: The name of the simulation, if provided.
#'   - `$sim_time_exceeded`: Whether the simulation stopped because the final simulation time was reached.
#'   - `$all_zero_state`: Whether an extinction has occurred.
#'   - `$negative_state`: Whether a negative state has occurred. If an SSA method other than `ssa_etl()` is used, this indicates a mistake in the provided reaction effects.
#'   - `$all_zero_propensity`: Whether the simulation stopped because all propensity values are zero.
#'   - `$negative_propensity`: Whether a negative propensity value has occurred. If so, there is likely a mistake in the provided reaction propensity functions.
#'   - `$walltime_exceeded`: Whether the simulation stopped because the maximum execution time has been reached.
#'   - `$walltime_elapsed`: The duration of the simulation.
#'   - `$num_steps`: The number of steps performed.
#'   - `$dtime_mean`: The mean time increment per step.
#'   - `$dtime_sd`: THe standard deviation of time increments.
#'   - `$firings_mean`: The mean number of firings per step.
#'   - `$firings_sd`: The standard deviation of the number of firings.
#'
#' @examples
#' \donttest{
#' initial_state <- c(prey = 1000, predators = 1000)
#' params <- c(c1 = 10, c2 = 0.01, c3 = 10)
#' reactions <- list(
#'   #        propensity function     effects                       name for reaction
#'   reaction(~c1 * prey,             c(prey = +1),                 "prey_up"),
#'   reaction(~c2 * prey * predators, c(prey = -1, predators = +1), "predation"),
#'   reaction(~c3 * predators,        c(predators = -1),            "pred_down")
#' )
#'
#' out <-
#'   ssa(
#'     initial_state = initial_state,
#'     reactions = reactions,
#'     params = params,
#'     method = ssa_exact(),
#'     final_time = 5,
#'     census_interval = .001,
#'     verbose = TRUE
#'   )
#'
#' library(ggplot2)
#' autoplot(out)
#' }
#'
#' @export
#'
#' @importFrom methods is
#' @importFrom dynutils is_sparse inherit_default_params
#' @importFrom Matrix Matrix
#' @importFrom purrr is_scalar_double is_scalar_logical is_scalar_integer is_scalar_character
ssa <- dynutils::inherit_default_params(
  list(create_simulation),
  function(
    initial_state,
    reactions,
    final_time,
    params,
    method = ssa_exact(),
    census_interval,
    stop_on_neg_state,
    max_walltime,
    log_propensity,
    log_firings,
    log_buffer,
    verbose,
    console_interval,
    sim_name
  ) {

    # check parameters
    assert_that(
      # initial state
      is.numeric(initial_state),
      !is.null(names(initial_state)),
      !any(c("state", "time", "params") %in% names(initial_state)),

      # params
      is.numeric(params),
      length(params) == 0 || !is.null(names(params)),
      !any(c("state", "time", "params") %in% names(params)),
      !any(duplicated(c(names(initial_state), names(params)))),

      # method
      is(method, "SSA_method"),

      # vector arguments
      is_scalar_double(final_time), final_time >= 0,
      is_scalar_double(census_interval), census_interval >= 0,
      is_scalar_double(max_walltime), max_walltime >= 0,
      is_scalar_double(console_interval), console_interval >= 0,
      is_scalar_logical(log_propensity),
      is_scalar_logical(log_firings),
      is_scalar_logical(log_buffer),
      is_scalar_logical(verbose),
      is_scalar_character(sim_name)
    )

    # compile propensity functions if this has not been done already
    compiled_reactions <-
      if (is.list(reactions) && !is(reactions, "SSA_reactions")) {
        compile_reactions(
          reactions = reactions,
          state_ids = names(initial_state),
          params = params
        )
      } else {
        reactions
      }

    assert_that(is(compiled_reactions, "SSA_reactions"))

    # create new simulation object
    sim <- create_simulation(
      compiled_reactions = compiled_reactions,
      params = params,
      method_ptr = method$factory(),
      initial_state = initial_state,
      census_interval = census_interval,
      log_propensity = log_propensity,
      log_firings = log_firings,
      log_buffer = log_buffer,
      stop_on_neg_state = stop_on_neg_state,
      final_time = final_time,
      max_walltime = max_walltime,
      sim_name = sim_name,
      verbose = verbose,
      console_interval = console_interval
    )

    # run simulation
    sim$run()

    output <- list(
      time = sim$output_time,
      state = sim$output_state,
      propensity = sim$output_propensity,
      firings = sim$output_firings,
      buffer = sim$output_buffer,
      stats = sim$get_statistics()
    )

    rm(sim)

    # set colnames of objects
    colnames(output$state) <- names(initial_state)
    if (log_propensity) {
      colnames(output$propensity) <- compiled_reactions$reaction_ids
    }
    if (log_buffer) {
      colnames(output$buffer) <- compiled_reactions$buffer_ids
    }
    if (log_firings) {
      colnames(output$firings) <- compiled_reactions$reaction_ids
    }

    class(output) <- c("ssa", "list")

    output
  }
)

File: src/ssa_simulation.cpp
#include <Rcpp.h>
#include <math.h>
#include <Rcpp/Benchmark/Timer.h>

#include "ssa_method.h"
#include "utils.h"

using namespace Rcpp;

typedef void (*TR_FUN)(const NumericVector&, const NumericVector&, const double, NumericVector&, NumericVector&);

class SSA_simulation {
public:
  SSA_simulation() {}

  ~SSA_simulation() {
    // Rcout << "FREE WILLY" << std::endl;
    if (prop_funs) {
      delete[] prop_funs;
    }
  }

  void initialise(
    // propensity functions
    List propensity_funs_,
    NumericVector& params_,
    int buffer_size_,
    // method
    SEXP ssa_method_,
    // state
    NumericVector& initial_state_,
    // state change matrix
    IntegerVector& nu_i_,
    IntegerVector& nu_p_,
    IntegerVector& nu_x_,
    // output parameters
    double census_interval_,
    bool log_propensity_,
    bool log_firings_,
    bool log_buffer_,
    // stopping conditions
    bool stop_on_neg_state_,
    double final_time_,
    double max_walltime_,
    // meta information
    std::string sim_name_,
    bool verbose_,
    double console_interval_
  ) {
    // process prop funs
    num_functions = propensity_funs_.size();
    prop_funs = new TR_FUN[num_functions];
    for (int i = 0; i < num_functions; i++) {
      prop_funs[i] = *XPtr<TR_FUN>(as<SEXP>(propensity_funs_[i]));
    }
    params = params_;
    buffer = NumericVector(buffer_size_);

    // process algorithm
    if (!Rf_isNull(ssa_method_)) {
      ssa_alg = XPtr<SSA_method>(ssa_method_);
    }

    // process state
    initial_state = initial_state_;
    state = NumericVector(initial_state.size());
    dstate = NumericVector(initial_state.size());

    // process state change matrix
    nu_i = nu_i_;
    nu_p = nu_p_;
    nu_x = nu_x_;

    propensity = NumericVector(nu_p.size() - 1);
    firings = NumericVector(propensity.size());
    dfirings = NumericVector(propensity.size());

    // process output parameters
    census_interval = census_interval_;
    log_propensity = log_propensity_;
    log_firings = log_firings_;
    log_buffer = log_buffer_;

    // process stopping conditions
    stop_on_neg_state = stop_on_neg_state_;
    final_time = final_time_;
    max_walltime = max_walltime_;

    // meta information
    sim_name = sim_name_;
    verbose = verbose_;
    console_interval = console_interval_;

    // initialise output structures
    output_nexti = 0;
    output_time = NumericVector(10);
    output_state = NumericMatrix(10, state.size());
    if (log_propensity) {
      output_propensity = NumericMatrix(10, propensity.size());
    } else {
      output_propensity = NumericMatrix(0, 0);
    }
    if (log_buffer) {
      output_buffer = NumericMatrix(10, buffer.size());
    } else {
      output_buffer = NumericMatrix(0, 0);
    }
    if (log_firings) {
      output_firings = NumericMatrix(10, firings.size());
    } else {
      output_firings = NumericMatrix(0, 0);
    }
  }

  void reset() {
    // reset output strunctures
    output_nexti = 0;
    resize_outputs(10, true);
    sim_time_nextcensus = census_interval;

    // initialise state data structures
    sim_time = 0;
    dtime = 0;
    std::copy(initial_state.begin(), initial_state.end(), state.begin());
    std::fill(dstate.begin(), dstate.end(), 0);
    std::fill(buffer.begin(), buffer.end(), 0);
    std::fill(firings.begin(), firings.end(), 0);
    std::fill(dfirings.begin(), dfirings.end(), 0);

    // statistics data structures
    num_steps = 0;
    dtime_mean = 0;
    dtime_sd = 0;
    firings_mean = 0;
    firings_sd = 0;
    walltime_elapsed = 0;

    // stopping conditions
    negative_state = false;
    negative_propensity = false;
    all_zero_propensity = false;
    all_zero_state = false;

    calculate_propensity();
    do_census();
  }

  void run() {
    // initialise walltime fields
    Timer timer;
    timer.step("start");

    nanotime_t walltime_start = timer.origin(),
      walltime_nextconsole = timer.origin(),
      walltime_nextinterrupt = timer.origin(),
      walltime_curr = timer.origin();

    // verbose
    if (verbose) {
      Rcout << "Running SSA " << ssa_alg->name << " with console output every " << console_interval << " seconds" << std::endl;
    }

    while (
        sim_time < final_time &&
          (walltime_curr - walltime_start) / 1000000000 < max_walltime &&
          !negative_propensity &&
          !all_zero_propensity &&
          (!negative_state || !stop_on_neg_state)
    )  {

      // check for interrupt
      if (walltime_nextinterrupt <= timer.now()) {
        checkUserInterrupt();
        walltime_nextinterrupt += 1000000000;
      }

      // print if so desired
      if (verbose && walltime_nextconsole <= walltime_curr) {
        Rcout << "walltime: " << (walltime_curr - walltime_start) << ", sim_time: " << sim_time << std::endl;
        walltime_nextconsole += console_interval * 1000000000;
      }

      // make a step
      make_step();

      // recalculate propensity
      calculate_propensity();

      // perform census if so desired
      if (sim_time_nextcensus <= sim_time) {
        sim_time_nextcensus += census_interval;
        do_census();
      }

      walltime_curr = timer.now();
    }

    // log end state if census_interval is set to inf
    if (std::isinf(census_interval)) {
      do_census();
    }

    // determine whether extinction has occurred
    all_zero_state = true;
    for (NumericVector::iterator i = state.begin(); i != state.end() && all_zero_state; i++) {
      if (*i > 0) {
        all_zero_state = false;
      }
    }

    // construct output
    nanotime_t walltime_end = timer.now();
    walltime_elapsed += walltime_end - walltime_start;

    // remove empty output slots
    resize_outputs(output_nexti, false);

    if (verbose) {
      Rcout << "SSA finished!" << std::endl;
    }
  }

  DataFrame get_statistics() {
    return DataFrame::create(
      _["method"] = ssa_alg->name,
      _["sim_name"] = sim_name,
      _["sim_time_exceeded"] = sim_time > final_time,
      _["all_zero_state"] = all_zero_state,
      _["negative_state"] = negative_state,
      _["all_zero_propensity"] = all_zero_propensity,
      _["negative_propensity"] = negative_propensity,
      _["walltime_exceeded"] = walltime_elapsed >= max_walltime,
      _["walltime_elapsed"] = walltime_elapsed / 1000000000.0,
      _["num_steps"] = num_steps,
      _["dtime_mean"] = dtime_mean,
      _["dtime_sd"] = dtime_sd,
      _["firings_mean"] = firings_mean,
      _["firings_sd"] = firings_sd
    );
  }

  void do_census() {
    // enlarge output objects, if needed
    if (output_nexti == output_time.size()) {
      resize_outputs(output_nexti * 2, false);
    }

    // copy state to output objects
    output_time[output_nexti] = sim_time;
    output_state(output_nexti, _) = state;
    if (output_propensity.nrow() > 0) {
      output_propensity(output_nexti, _) = propensity;
    }
    if (output_buffer.nrow() > 0) {
      output_buffer(output_nexti, _) = buffer;
    }
    if (output_firings.nrow() > 0) {
      output_firings(output_nexti, _) = firings;
      std::fill(firings.begin(), firings.end(), 0);
    }
    output_nexti++;
  }

  void calculate_propensity() {
    for (int i = 0; i < num_functions; i++) {
      prop_funs[i](state, params, sim_time, propensity, buffer);
    }

    // check whether all propensity functions are zero
    all_zero_propensity = true;
    for (NumericVector::iterator i = propensity.begin(); i != propensity.end(); ++i) {
      if (*i > 0) {
        all_zero_propensity = false;
      } else if (*i < 0) {
        negative_propensity = true;
      }
    }
  }

  void make_step() {
    // clear dtime, dstate and dfirings
    dtime = 0;
    std::fill(dstate.begin(), dstate.end(), 0);
    std::fill(dfirings.begin(), dfirings.end(), 0);

    // make a step
    ssa_alg->step(state, propensity, nu_i, nu_p, nu_x, &dtime, dstate, dfirings);

    num_steps++;
    sim_time += dtime;
    state += dstate;
    firings += dfirings;

    // update statistics
    int firings_sum = sum(dfirings);
    if (num_steps == 1) {
      dtime_sd = 0;
      firings_sd = 0;
    } else {
      dtime_sd = sqrt((num_steps - 2) / (num_steps - 1) * pow(dtime_sd, 2) + pow(dtime - dtime_mean, 2) / num_steps);
      firings_sd = sqrt((num_steps - 2) / (num_steps - 1) * pow(firings_sd, 2) + pow(firings_sum - firings_mean, 2) / num_steps);
    }
    dtime_mean = (dtime_mean * (num_steps - 1) + dtime) / num_steps;
    firings_mean = (firings_mean * (num_steps - 1) + firings_sum) / num_steps;

    // Check that no states are negative (can occur in some tau-leaping methods)
    for (NumericVector::iterator i = state.begin(); i != state.end(); ++i) {
      if (*i < 0) {
        if (!stop_on_neg_state) {
          *i = 0;
        }
        negative_state = true;
      }
    }
  }

  void resize_outputs(int num_rows, bool fill_zero) {
    output_time = resize_vector(output_time, num_rows, !fill_zero);
    output_state = resize_matrix(output_state, num_rows, output_state.ncol(), !fill_zero);
    if (output_propensity.nrow() > 0) {
      output_propensity = resize_matrix(output_propensity, num_rows, output_propensity.ncol(), !fill_zero);
    }
    if (output_buffer.nrow() > 0) {
      output_buffer = resize_matrix(output_buffer, num_rows, output_buffer.ncol(), !fill_zero);
    }
    if (output_firings.nrow() > 0) {
      output_firings = resize_matrix(output_firings, num_rows, output_firings.ncol(), !fill_zero);
    }

    if (fill_zero) {
      std::fill(output_time.begin(), output_time.end(), 0);
      std::fill(output_state.begin(), output_state.end(), 0);
      std::fill(output_propensity.begin(), output_propensity.end(), 0);
      std::fill(output_buffer.begin(), output_buffer.end(), 0);
      std::fill(output_firings.begin(), output_firings.end(), 0);
    }
  }

  template <typename T>
  T resize_vector(const T& x, int n, bool copy){
    int oldsize = x.size();
    if (n == oldsize) {
      return x;
    } else if (n < oldsize) {
      oldsize = n;
    }
    T y(n); // TODO: no init?
    if (copy) {
      for( int i = 0; i < oldsize; i++) {
        y[i] = x[i];
      }
    }
    return y;
  }

  template <typename T>
  T resize_matrix(const T& x, int nr, int nc, bool copy){
    int oldnr = x.nrow();
    int oldnc = x.ncol();
    if (nr == oldnr && nc == oldnc) {
      return x;
    }
    if (nr < oldnr) {
      oldnr = nr;
    }
    if (nc < oldnc) {
      oldnc = nc;
    }
    T y(nr, nc); // TODO: no init?
    if (copy) {
      for( int i = 0; i < oldnr; i++) {
        for (int j = 0; j < oldnc; j++) {
          y(i, j) = x(i, j);
        }
      }
    }
    return y;
  }

  int num_functions;
  TR_FUN *prop_funs;
  SSA_method *ssa_alg;
  NumericVector initial_state;
  NumericVector params;
  IntegerVector nu_i;
  IntegerVector nu_p;
  IntegerVector nu_x;

  // state data structures
  double sim_time;
  double dtime;
  NumericVector state;
  NumericVector dstate;
  NumericVector propensity;
  NumericVector buffer;
  NumericVector firings;
  NumericVector dfirings;

  // statistics data structures
  int num_steps;
  double dtime_mean;
  double dtime_sd;
  double firings_mean;
  double firings_sd;
  double walltime_elapsed;

  // log data structures
  int output_nexti;
  NumericVector output_time;
  NumericMatrix output_state;
  NumericMatrix output_propensity;
  NumericMatrix output_buffer;
  NumericMatrix output_firings;

  // output parameters
  double census_interval;
  double sim_time_nextcensus;
  bool log_propensity;
  bool log_firings;
  bool log_buffer;

  // stopping conditions
  bool all_zero_propensity;
  bool all_zero_state;
  bool negative_state;
  bool negative_propensity;
  bool stop_on_neg_state;
  double final_time;
  double max_walltime;

  // meta information
  std::string sim_name;
  bool verbose;
  double console_interval;
};


// export everything so it can be unit tested
RCPP_MODULE(gillespie) {
  class_<SSA_simulation>("SSA_simulation")
  .constructor()
  .method("initialise", &SSA_simulation::initialise)
  .method("run", &SSA_simulation::run)
  .method("reset", &SSA_simulation::reset)
  .method("get_statistics", &SSA_simulation::get_statistics)
  .method("do_census", &SSA_simulation::do_census)
  .method("calculate_propensity", &SSA_simulation::calculate_propensity)
  .method("make_step", &SSA_simulation::make_step)
  .method("resize_outputs", &SSA_simulation::resize_outputs)
  .field("initial_state", &SSA_simulation::initial_state)
  .field("params", &SSA_simulation::params)
  .field("nu_i", &SSA_simulation::nu_i)
  .field("nu_p", &SSA_simulation::nu_p)
  .field("nu_x", &SSA_simulation::nu_x)
  .field("sim_time", &SSA_simulation::sim_time)
  .field("dtime", &SSA_simulation::dtime)
  .field("state", &SSA_simulation::state)
  .field("dstate", &SSA_simulation::dstate)
  .field("propensity", &SSA_simulation::propensity)
  .field("buffer", &SSA_simulation::buffer)
  .field("firings", &SSA_simulation::firings)
  .field("dfirings", &SSA_simulation::dfirings)
  .field("num_steps", &SSA_simulation::num_steps)
  .field("dtime_mean", &SSA_simulation::dtime_mean)
  .field("dtime_sd", &SSA_simulation::dtime_sd)
  .field("firings_mean", &SSA_simulation::firings_mean)
  .field("firings_sd", &SSA_simulation::firings_sd)
  .field("output_nexti", &SSA_simulation::output_nexti)
  .field("output_time", &SSA_simulation::output_time)
  .field("output_state", &SSA_simulation::output_state)
  .field("output_propensity", &SSA_simulation::output_propensity)
  .field("output_buffer", &SSA_simulation::output_buffer)
  .field("output_firings", &SSA_simulation::output_firings)
  .field("census_interval", &SSA_simulation::census_interval)
  .field("log_propensity", &SSA_simulation::log_propensity)
  .field("log_firings", &SSA_simulation::log_firings)
  .field("log_buffer", &SSA_simulation::log_buffer)
  .field("all_zero_propensity", &SSA_simulation::all_zero_propensity)
  .field("all_zero_state", &SSA_simulation::all_zero_state)
  .field("negative_propensity", &SSA_simulation::negative_propensity)
  .field("negative_state", &SSA_simulation::negative_state)
  .field("stop_on_neg_state", &SSA_simulation::stop_on_neg_state)
  .field("final_time", &SSA_simulation::final_time)
  .field("max_walltime", &SSA_simulation::max_walltime)
  .field("sim_name", &SSA_simulation::sim_name)
  .field("verbose", &SSA_simulation::verbose)
  .field("console_interval", &SSA_simulation::console_interval)
  ;
}
